app/jobs/download_listing_image_job.rb

```diff
@@ -13,44 +13,56 @@ def perform
     # It's doing network operations, so I guess it can also throw.
     #
 
-    find_listing_image(listing_image_id).and_then { |listing_image|
+    listing_image.and_then { |listing_image|
 
-      # Download the original image
       begin
+        # Download the original image
         listing_image.image = URI.parse(url)
+        # Save the image, create delayed jobs for processing, update the download status
+        listing_image.update_attribute(:image_downloaded, true)
         Result::Success.new(listing_image)
       rescue StandardError => e
         Result::Error.new(e)
       end
 
-    }.and_then { |listing_image|
+    }.on_error { |error_msg, ex|
+      logger.error(error_msg, :listing_image_download_failed)
 
-      # Save the image, create delayed jobs for processing, update the download status
-      begin
-        listing_image.update_attribute(:image_downloaded, true)
-        Result::Success.new(listing_image)
-      rescue StandardError => e
-        Result::Error.new(e)
-      end
+      raise ex # Reraise the exception to mark the delayed job failed
+    }
+  end
 
+  def failure
+    listing_image.on_success { |listing_image|
+      listing_image.update_column(:errored, true) # Have to use `update_column` here because `update_attribute` runs the before save hooks
```

Comment:
I was wondering if instead of errored just being a boolean it could contain the error reason as symbol but I'm not sure if the exception data is available here anymore?